### PR TITLE
Add return address stack location to context backtrace #1288

### PR DESCRIPTION
**Pull request by dmur1**
* Add pwndbg.gdblib.find_addr_on_stack

* Change coloration of context backtrace to use pwnlib.color.memory

* Remove now defunct address color config from context

* Update test_context_backtrace_show_proper_symbol_names regex

### CODE REVIEW

File: pwndbg/commands/context.py [hunk 1]
```diff
@@ -41,7 +41,6 @@
     "backtrace",
     [
         ColorParamSpec("prefix", "none", "color for prefix of current backtrace label"),
-        ColorParamSpec("address", "none", "color for backtrace (address)"),
         ColorParamSpec("symbol", "none", "color for backtrace (symbol)"),
```
Comment (dmur1) [inline pwndbg/commands/context.py:44]: i wonder if we should use CODE coloration for symbols too? as is the case in the register output for example:

image

File: pwndbg/gdblib/stack.py [hunk 2]
```diff
@@ -136,3 +136,27 @@ def is_executable() -> bool:
             nx = True
 
     return not nx
+
+
+def find_addr_on_stack(addr):
```
Comment (dmur1) [inline pwndbg/gdblib/stack.py]: this method is basically copied from here https://github.com/pwndbg/pwndbg/blob/dev/pwndbg/commands/stack.py#L26

i wonder if both should use this centralised approach?

File: tests/gdb-tests/tests/test_context_commands.py [hunk 3]
```diff
@@ -201,21 +201,19 @@ def test_context_backtrace_show_proper_symbol_names(start_binary):
         == "─────────────────────────────────[ BACKTRACE ]──────────────────────────────────"
     )
 
-    assert re.match(r".*0   0x[0-9a-f]+ A::foo\(int, int\)", backtrace[2])
```
Comment (dmur1) [inline tests/gdb-tests/tests/test_context_commands.py:204]: how important is it to tightly fit this regex to the entire output in a test that is mainly of proper symbol names?

File: pwndbg/gdblib/stack.py (hunk 2 above)
Comment (disconnect3d) [reply pwndbg/gdblib/stack.py]: Yeah I think it makes sense to centralize it. Also I believe there's a small edge case bug, it should be:
```diff
-while addresses and start < sp < stop:
+while addresses and start <= sp < stop:
```

File: tests/gdb-tests/tests/test_context_commands.py (hunk 3 above)
Comment (disconnect3d) [reply tests/gdb-tests/tests/test_context_commands.py:204]: The more accurate we are the less regressions we will have :). I am fine with more relaxed assertions: its better to have some tests than none and its hard to test for everything.

File: pwndbg/gdblib/stack.py (hunk 2 above)
Comment (dmur1) [reply pwndbg/gdblib/stack.py]: i'll make another issue to centralise that in another change i think ( and i'll fix that edge case and note it in the centralisation issue )

File: pwndbg/commands/context.py [hunk 4]
```diff
@@ -801,11 +800,14 @@ def context_backtrace(with_banner=True, target=sys.stdout, width=None):
     bt_prefix = "%s" % pwndbg.gdblib.config.backtrace_prefix
     while True:
         prefix = bt_prefix if frame == this_frame else " " * len(bt_prefix)
-        prefix = f" {c.prefix(prefix)}"
-        addrsz = c.address(pwndbg.ui.addrsz(frame.pc()))
-        symbol = c.symbol(pwndbg.gdblib.symbol.get(int(frame.pc())))
+        prefix = " %s" % c.prefix(prefix)
```
Comment (gsingh93) [inline pwndbg/commands/context.py]: Can you leave this as an f-string? I'm trying to remove all instances of `%s` from the codebase.

File: pwndbg/commands/context.py [hunk 5]
```diff
@@ -801,11 +800,14 @@ def context_backtrace(with_banner=True, target=sys.stdout, width=None):
...
+        frame_pc = frame.pc()
+        stackaddr = M.get((pwndbg.gdblib.stack.find_addr_on_stack(frame_pc)))
+        addrsz = "(" + stackaddr + ") " + M.get(frame_pc)
```
Comment (gsingh93) [inline pwndbg/commands/context.py]: Use an `f-string` here to avoid the concatenation.

File: pwndbg/commands/stack.py [hunk 6]
```diff
@@ -26,7 +26,7 @@ def retaddr() -> None:
     # Find all of them on the stack
     start = stack.vaddr
     stop = start + stack.memsz
-    while addresses and start < sp < stop:
+    while addresses and start <= sp < stop:
```
Comment (gsingh93) [inline pwndbg/commands/stack.py:29]: I guess this is to cover the case when the return address is pushed to the stack but we haven't pushed a frame pointer yet, so `$rsp` points to the return address and without the `<=` change we wouldn't print it (cc: @disconnect3d )? If so, it would be good to add a test for this.

File: pwndbg/gdblib/stack.py (hunk 2 above)
Comment (gsingh93) [inline pwndbg/gdblib/stack.py]: Can you add some type hints here?

File: pwndbg/commands/context.py [hunk 7]
```diff
@@ -801,11 +800,14 @@ def context_backtrace(with_banner=True, target=sys.stdout, width=None):
     bt_prefix = "%s" % pwndbg.gdblib.config.backtrace_prefix
     while True:
         prefix = bt_prefix if frame == this_frame else " " * len(bt_prefix)
-        prefix = f" {c.prefix(prefix)}"
-        addrsz = c.address(pwndbg.ui.addrsz(frame.pc()))
-        symbol = c.symbol(pwndbg.gdblib.symbol.get(int(frame.pc())))
+        prefix = " %s" % c.prefix(prefix)
+        frame_pc = frame.pc()
+        stackaddr = M.get((pwndbg.gdblib.stack.find_addr_on_stack(frame_pc)))
```
Comment (gsingh93) [inline pwndbg/commands/context.py]: It's a little hacky, but to deal with the alignment issue you brought up [here](https://github.com/pwndbg/pwndbg/issues/1288#issuecomment-1595693469) we can call the [`M.get`](https://github.com/pwndbg/pwndbg/blob/dev/pwndbg/color/memory.py#L19) function with the optional `text` parameter. We can convert the address to a hex string that's already padded to a width of `pwndbg.gdblib.arch.ptrsize`, use that as the `text` argument, and let `M.get` color it properly.

Also nit: seems like there are extra parenthesis here you can remove.

**Review by gsingh93 — CHANGES_REQUESTED** [review]
Nice work overall, just needs a few minor changes.

File: pwndbg/commands/context.py (hunk 1 above)
Comment (disconnect3d) [reply pwndbg/commands/context.py:44]: If we display symbols, probably yes.

**Comment by disconnect3d** [comment]
@dmur1 can u show a screenshot before and after?

File: pwndbg/commands/context.py (hunk 5 above)
Comment (dmur1) [reply pwndbg/commands/context.py]: done

File: pwndbg/commands/stack.py (hunk 6 above)
Comment (dmur1) [reply pwndbg/commands/stack.py:29]: pinging @disconnect3d about this one again because i'm not totally sure

File: pwndbg/gdblib/stack.py (hunk 2 above)
Comment (dmur1) [reply pwndbg/gdblib/stack.py]: done

File: pwndbg/commands/context.py (hunk 7 above)
Comment (dmur1) [reply pwndbg/commands/context.py]: done

**Review by dmur1 — COMMENTED** [review]
take a look at the newest stuff!

**Comment by dmur1** [comment]
> @dmur1 can u show a screenshot before and after?

before:
image

after:
image

File: pwndbg/gdblib/stack.py [hunk 8]
```diff
@@ -136,3 +136,27 @@ def is_executable() -> bool:
             nx = True
 
     return not nx
+
+
+def find_addr_on_stack(addr) -> int:
```
Comment (disconnect3d) [inline pwndbg/gdblib/stack.py:141]: I am a bit worried that this may be (way) too slow.

* We are reading the whole stack memory: what if stack is super big?
* What if SP does not point to mapped memory? This will crash
* This can be optimized out in two ways:
1) We could leverage the existing search functionality we have which will do a find using GDB's API which is written in C and so should be faster
2) We could read the whole stack memory and use simple `.find(...)` to find the offset of the bytes of the `addr` value

The 2) will rather be slower than 1) but if we read the stack memory just once, then we can find the fetched bytes multiple times to look for different return addresses --- which would probably be faster overall?

Comment (disconnect3d) [reply pwndbg/gdblib/stack.py:141]: Given this I am not sure if we really want to look for those values on the stack... we are making context slower this way. We can however make this configurable.

@gsingh93 thoughts?

File: pwndbg/commands/context.py [hunk 9]
```diff
@@ -802,10 +801,15 @@ def context_backtrace(with_banner=True, target=sys.stdout, width=None):
     while True:
         prefix = bt_prefix if frame == this_frame else " " * len(bt_prefix)
         prefix = f" {c.prefix(prefix)}"
-        addrsz = c.address(pwndbg.ui.addrsz(frame.pc()))
-        symbol = c.symbol(pwndbg.gdblib.symbol.get(int(frame.pc())))
+        frame_pc = frame.pc()
+        frame_pc_on_stack = pwndbg.gdblib.stack.find_addr_on_stack(frame_pc)
+        padded_text = "0x" + hex(frame_pc_on_stack)[2:].zfill(pwndbg.gdblib.arch.ptrsize * 2)
```
Comment (disconnect3d) [inline pwndbg/commands/context.py]: `hex(...)` returns a `0x...` string, so we don't need to `[2:]` it and prepend a `0x` string

Also nitpick, but we fetch and compute `pwndbg.gdblib.arch.ptrsize * 2` in each loop iteration :<

File: pwndbg/commands/context.py [hunk 10]
```diff
@@ -802,10 +801,15 @@ def context_backtrace(with_banner=True, target=sys.stdout, width=None):
...
+        stackaddr = M.get(frame_pc_on_stack, padded_text)
+        addrsz = "(" + stackaddr + ") " + M.get(frame_pc)
```
Comment (disconnect3d) [inline pwndbg/commands/context.py]: Can we use f-string literals here and wherever we can?

**Comment by disconnect3d** [comment]
> > @dmur1 can u show a screenshot before and after?
> 
> before: image
> 
> after: image

Random thoughts about the display:
1) I don't like the `0x<zeros>` prepended
2) Maybe we should display `<sp register>+0x<OFFSET>` instead of whole addresses?

...or maybe this should be configurable, ugh. cc: @gsingh93

File: pwndbg/commands/context.py (hunk 9 above)
Comment (dmur1) [reply pwndbg/commands/context.py]: >>> hex(1234).zfill(8)
'0000x4d2'

i was trying to avoid this - perhaps theres a way to zfill an already 0x prefixed string?

@disconnect3d

Comment (dmur1) [reply pwndbg/commands/context.py]: will hoist the computation out of the loop

**Comment by dmur1** [comment]
> > > @dmur1 can u show a screenshot before and after?
> > 
> > 
> > before: image
> > after: image
> 
> Random thoughts about the display:
> 
> 1. I don't like the `0x<zeros>` prepended
> 2. Maybe we should display `<sp register>+0x<OFFSET>` instead of whole addresses?
> 
> ...or maybe this should be configurable, ugh. cc: @gsingh93

i have to admit i'm also not a fan especially as it contrasts with the other aligned addresses in the other context sections

File: pwndbg/commands/context.py (hunk 9 above)
Comment (disconnect3d) [reply pwndbg/commands/context.py]: I'm not sure about this zfill. Like, do we really want to display those zeroes? :P They are redundant

**Comment by dmur1** [comment]
bumping this - i'm still not sure i like it what do you folks think?

**Comment by disconnect3d** [comment]
I think prefixing with additional zeroes is not great. What else do u ask about? :P

**Comment by dmur1** [comment]
> I think prefixing with additional zeroes is not great. What else do u ask about? :P

so maybe if just padded as much as was needed to keep everything aligned?

**Comment by disconnect3d** [comment]
> > I think prefixing with additional zeroes is not great. What else do u ask about? :P
> 
> so maybe if just padded as much as was needed to keep everything aligned?

Either this, or no padding at all?

**Comment by disconnect3d** [comment]
(This now needs to be rebased and conflicts needs to be fixed :()

**Comment by disconnect3d** [comment]
Closing it for now; I'm still up for accepting such feature in case anyone wants to add it
